Perlang.Interpreter/Typing/TypeValidator.cs

```diff
@@ -110,15 +116,26 @@ public void Resolve(IList<Stmt> statements)
                 Visit(statements);
             }
 
-            public void Resolve(Expr expr)
-            {
-                Visit(expr);
-            }
-
             //
             // Expr visitors
             //
 
+            public override VoidObject VisitAssignExpr(Expr.Assign expr)
+            {
+                base.VisitAssignExpr(expr);
+
+                // Letting the type be inferred in an assignment expression is important to make constructs like
+                // "var i = 100; var j = i+= 2;" work correctly. This is indeed an odd way of writing code, but as long
+                // as += is an expression and not a statement, we need to have predictable semantics for cases like
+                // this.
+                if (!expr.TypeReference.IsResolved && expr.Value.TypeReference.IsResolved)
+                {
+                    expr.TypeReference.ClrType = expr.Value.TypeReference.ClrType;
+                }
+
+                return VoidObject.Void;
+            }
+
```

Comment:
This is the gotcha I mentioned in the PR description. It took me a bit of time to track this one down. The code would _work_ for many circumstances, but not in cases like `var j = i += 2`. The problem was that the type inference in the `i + 2` (implicit binary expression) was never properly propagated to the `i = i + 2` parent expression - the `Expr.Assign` instance referred to here.

The exact problem we saw here (a binary expression inside an assignment expression) was not necessarily present before the `+=` and `-=` implementation introduced in this PR; such expressions would perhaps never occur before these changes. But still, constructs like this would fail with commits like 3263a42 and 70e5ea5:

```perlang
var i = 100;
var k = (i = i + 1);
```

The above gives me an `Internal compiler error: var k initializer Perlang.Expr+Grouping inference has not been attempted` exception. With this change, the above exception is gone. :rocket: 

(The integration tests introduced by this MR highlighted the problem, so I think we are fine coverage-wise for now.)